pandas/core/case_when.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import annotations
+
+from typing import Any
+import warnings
+
+from pandas.util._exceptions import find_stack_level
+
+import pandas as pd
+import pandas.core.common as com
+
+
+def warn_and_override_index(series, series_type, index):
+    warnings.warn(
+        f"Series {series_type} will be reindexed to match obj index.",
+        UserWarning,
+        stacklevel=find_stack_level(),
+    )
+    return pd.Series(series.values, index=index)
+
+
+def case_when(obj: pd.DataFrame | pd.Series, *args, default: Any) -> pd.Series:
+    """
+    Returns a Series based on multiple conditions assignment.
+
+    This is useful when you want to assign a column based on multiple conditions.
+    Uses `Series.mask` to perform the assignment.
+
+    If multiple conditions are met, the value of the first condition is taken.
+
+    The returned Series have the same index as `obj` as mainly follows dtype
+    of `default`.
+
+    Parameters
+    ----------
+    obj : Dataframe or Series on which the conditions will be applied.
```

Comment:
The first line should just be the dtype; a description of the argument should appear on the subsequent line.